File: patterns/golden_cross.py
# trade-signals-web/patterns/golden_cross.py

import pandas as pd
from .base_pattern import BasePattern

class GoldenCrossPattern(BasePattern):
    """Золотой крест (50/200 MA)"""
    
    def __init__(self):
        super().__init__("Золотой крест (50/200)", confidence="high")
    
    def get_min_candles(self) -> int:
        return 200
    
    def detect(self, df: pd.DataFrame) -> dict | None:
        df['sma50'] = df['close'].rolling(window=50).mean()
        df['sma200'] = df['close'].rolling(window=200).mean()
        
        if len(df) < 2:
            return None
        
        prev = df.iloc[-2]
        curr = df.iloc[-1]
        
        if (pd.notna(prev['sma50']) and pd.notna(prev['sma200']) and 
            pd.notna(curr['sma50']) and pd.notna(curr['sma200'])):
            if prev['sma50'] <= prev['sma200'] and curr['sma50'] > curr['sma200']:
                return {
                    "signal": "BUY",
                    "description": "50MA пересекла 200MA снизу вверх",
                    "confidence": self.confidence
                }
        
        return None

File: js/patterns/golden_cross.js
// js/patterns/golden_cross.js

import { calculateEMA } from '../indicators/ema.js';

export class GoldenCrossPattern {
    name = 'Золотой крест (50/200)';
    confidence = 'high';

    detect(candles) {
        if (candles.length < 200) return null;

        const closes = candles.map(c => c.close);
        const ema50 = calculateEMA(closes, 50);
        const ema200 = calculateEMA(closes, 200);

        if (ema50.length < 2 || ema200.length < 2) return null;

        const prev50 = ema50[ema50.length - 2];
        const curr50 = ema50[ema50.length - 1];
        const prev200 = ema200[ema200.length - 2];
        const curr200 = ema200[ema200.length - 1];

        if (prev50 <= prev200 && curr50 > curr200) {
            return {
                signal: 'BUY',
                description: '50MA пересекла 200MA снизу вверх',
                confidence: this.confidence
            };
        }

        return null;
    }
}
